kvserver: add merged `SSTTimestampToRequestTimestamp` param for `AddSSTable`

erikgrinaker opened the pull request:
**hlc: add `Timestamp.IsSet()`**

This patch adds `Timestamp.IsSet()`, which is the inverse of
`IsEmpty()`. In complex boolean expressions, `IsSet()` is significantly
easier to parse than `!IsEmpty()`, which can avoid mistakes.

Release note: None

**testutils/sstutil: tweak `MakeSST()` parameters**

This makes `testing.T` the first parameter, as is customary, and
replaces the unnecessary `Context` parameter with a background context
as this is only used for logging.

Release note: None

**kvserver: add merged `UpdateToRequestTimestamp` param for `AddSSTable`**

This merges the `AddSSTable` parameters `WriteAtRequestTimestamp` and
`SSTTimestamp` into a new `SSTTimestampToRequestTimestamp` parameter.
This parameter specifies the MVCC timestamp of all existing timestamps
in the SST, which will be rewritten to the request timestamp if they
differ (e.g. if the request gets pushed).

Both of the replaced parameters are new in 22.1, so this does not
require a version gate and allows changing the type of one of the
existing Protobuf fields.

Release note: None

erikgrinaker committed:
hlc: add `Timestamp.IsSet()`

This patch adds `Timestamp.IsSet()`, which is the inverse of
`IsEmpty()`. In complex boolean expressions, `IsSet()` is significantly
easier to parse than `!IsEmpty()`, which can avoid mistakes.

Release note: None

erikgrinaker committed:
kvserver: add merged `SSTTimestampToRequestTimestamp` param for `AddSSTable`

This merges the `AddSSTable` parameters `WriteAtRequestTimestamp` and
`SSTTimestamp` into a new `SSTTimestampToRequestTimestamp` parameter.
This parameter specifies the MVCC timestamp of all existing timestamps
in the SST, which will be rewritten to the request timestamp if they
differ (e.g. if the request gets pushed).

Both of the replaced parameters are new in 22.1, so this does not
require a version gate and allows changing the type of one of the
existing Protobuf fields.

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/76110)
<!-- Reviewable:end -->

sumeerbhola reviewed (commented):
> Reviewed 1 of 12 files at r2.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/76110)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @dt and @erikgrinaker)

___
*[pkg/roachpb/api.proto, line 1645 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/76110#-MvPlnBS20w_wj6C0s0-:-MvPlnBS20w_wj6C0s00:b-od5ffi) ([raw file](https://github.com/cockroachdb/cockroach/blob/193224eb7fee98b0b1110ff61e4e8da4e76c98f5/pkg/roachpb/api.proto#L1645)):*
> ```protobuf
>   bytes data = 2;
> 
>   // UpdateToRequestTimestamp updates all MVCC timestamps in the SST from the
> ```

(drive by comment. And the confusion may be limited to me so feel free to ignore.)

I found the "UpdateToRequestTimestamp updates all" and "parameter timestamp" phrasing a bit confusing. It also seemed that this was the timestamp to update to, when it is actually the timestamp in the sstable. How about something like:

// CurrentSSTableTimestampToRewrite when non-empty serves the following purpose:
// - It is the guaranteed timestamp of all entries in the sstable. This implies that the sstable only contains versioned keys.
// - It tells the proposal evaluator that if the request timestamp is not equal to this CurrentSSTableTimestampToRewrite, the evaluator must rewrite the sstable to the request timestamp.


<!-- Sent from Reviewable.io -->

erikgrinaker reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/76110)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @dt and @sumeerbhola)

___
*[pkg/roachpb/api.proto, line 1645 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/76110#-MvPlnBS20w_wj6C0s0-:-MvQ1TxO3bX69JGxlte6:bc5gv2f) ([raw file](https://github.com/cockroachdb/cockroach/blob/193224eb7fee98b0b1110ff61e4e8da4e76c98f5/pkg/roachpb/api.proto#L1645)):*
<details><summary><i>Previously, sumeerbhola wrote…</i></summary><blockquote>

(drive by comment. And the confusion may be limited to me so feel free to ignore.)

I found the "UpdateToRequestTimestamp updates all" and "parameter timestamp" phrasing a bit confusing. It also seemed that this was the timestamp to update to, when it is actually the timestamp in the sstable. How about something like:

// CurrentSSTableTimestampToRewrite when non-empty serves the following purpose:
// - It is the guaranteed timestamp of all entries in the sstable. This implies that the sstable only contains versioned keys.
// - It tells the proposal evaluator that if the request timestamp is not equal to this CurrentSSTableTimestampToRewrite, the evaluator must rewrite the sstable to the request timestamp.

</blockquote></details>

I like the concrete phrasing of the parameter's purpose. I don't love the name though. But names are hard. Maybe `SSTTimestampToRequestTimestamp`?


<!-- Sent from Reviewable.io -->

sumeerbhola reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/76110)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @dt and @erikgrinaker)

___
*[pkg/roachpb/api.proto, line 1645 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/76110#-MvPlnBS20w_wj6C0s0-:-Mv_OiHmEYomakZ-mn_D:b-h0tgb3) ([raw file](https://github.com/cockroachdb/cockroach/blob/193224eb7fee98b0b1110ff61e4e8da4e76c98f5/pkg/roachpb/api.proto#L1645)):*
<details><summary><i>Previously, erikgrinaker (Erik Grinaker) wrote…</i></summary><blockquote>

I like the concrete phrasing of the parameter's purpose. I don't love the name though. But names are hard. Maybe `SSTTimestampToRequestTimestamp`?
</blockquote></details>

I'm fine with whatever name you come up with


<!-- Sent from Reviewable.io -->

erikgrinaker reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/76110)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @dt and @sumeerbhola)

___
*[pkg/roachpb/api.proto, line 1645 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/76110#-MvPlnBS20w_wj6C0s0-:-MveXkfs2BfF2EjZ-6Gx:b-c2k38s) ([raw file](https://github.com/cockroachdb/cockroach/blob/193224eb7fee98b0b1110ff61e4e8da4e76c98f5/pkg/roachpb/api.proto#L1645)):*
<details><summary><i>Previously, sumeerbhola wrote…</i></summary><blockquote>

I'm fine with whatever name you come up with
</blockquote></details>

Renamed the parameter and reworded the comment.


<!-- Sent from Reviewable.io -->

erikgrinaker commented:
Pebble has been bumped, so this is good to go now.

erikgrinaker committed:
testutils/sstutil: tweak `MakeSST()` parameters

This makes `testing.T` the first parameter, as is customary, and
replaces the unnecessary `Context` parameter with a background context
as this is only used for logging.

Release note: None

aliher1911 commented on pkg/storage/sst.go:
```diff
@@ -263,11 +272,10 @@ func UpdateSSTTimestamps(
+		if key.Timestamp != from {
+			return nil, errors.Errorf("unexpected timestamp %s (expected %s) for key %s",
+				key.Timestamp, from, key.Key)
 		}
```
For my knowledge, why did we remove tombstone check, is it because we already verify that in assertSSTContents and there is nothing wrong at rewriting tombstone timestamp from "sst validity" perspective?

erikgrinaker replied:
This check already covers it. A tombstone has an empty timestamp, and we reject empty `from` timestamps at the start of this function, so a tombstone's timestamp would not equal `from` and therefore trigger this error. And yes, we also check it in `assertSSTContents`.

erikgrinaker replied:
Sorry, I was thinking about inline values, which we also used to check here.

You're right, we don't check tombstones here (empty value), because that doesn't have anything to do with `UpdateSSTTimestamps()` -- the job of this function is just to rewrite timestamps, and it doesn't care about the values. We do general `AddSSTable` validation in `assertSSTContents`.

aliher1911 commented on pkg/roachpb/api.proto:
```diff
@@ -1643,11 +1642,12 @@ message AddSSTableRequest {
-  // comply with the timestamp cache and closed timestamp. It also causes the
-  // AddSSTable to be emitted via the range feed, since it respects the closed
-  // timestamp.
+  // SSTTimestampToRequestTimestamp gives the timestamp used for all MVCC
```
nit: I see we had some discussion on this comment already, but it is still not ideal. SSTTimestampToRequestTimestamp doesn't rewrite SST itself, is it? Maybe something like:

SSTTimestampToRequestTimestamp is a timestamp used for all MVCC keys in the provided SST. If this timestamp is different from request timestamp (e.g. if the request gets pushed) all timestamps in SST will be rewritten by request timestamp at the time of request processing. This ensures...

erikgrinaker replied:
Yeah, that's better. Updated, with some tweaks.

aliher1911 commented on pkg/kv/kvserver/batcheval/cmd_add_sstable.go:
```diff
@@ -38,10 +38,11 @@ func init() {
 }
 
-var addSSTableRewriteConcurrency = settings.RegisterIntSetting(
+// AddSSTableRewriteConcurrency sets the concurrency of a single SST rewrite.
```
nit: I can see that concurrency term came from the pebble code, but isn't that parallelism here? We don't contend for resources and we rewrite multiple blocks in parallel. It trickles down all the way to seams with pebble, but that could be changes later as we please while settings are more user facing. I'll leave it to you if that's worth the change.

erikgrinaker replied:
I think for our purposes the terms are mostly interchangeable. Whether it's truly parallel will depend on e.g. the Go scheduler, OS scheduler, hardware, and workload (setting it to 8 on a single-core machine will not make it parallel). We use both terms in our existing settings, e.g.:

* `kv.bulk_io_write.restore_node_concurrency`
* `kv.rangefeed.max_scan_parallelism`

It's sort of unrelated to this PR in any case, so I'll leave it as-is.

aliher1911 replied:
I think parallel vs concurrent is more intent rather than execution. With parallel you want you computation to be split and processed independently to get benefit from multiple cores, otherwise there's no point. While concurrent is broader where we want all components run at the same time e.g. multitasking will give lower p90 latency v.s. executing sequentially on parallel requests.

aliher1911 reviewed (approved):
Some nits about naming and comments. Otherwise seems good.

erikgrinaker commented:
TFTR!

bors r=aliher1911
